File: routes/flow/done.js
var cates = require("./models/cates.js");
var goods = require("./models/goods.js");
var ordgoods = require("./models/ordgoods.js");
var ordinfos = require("./models/ordinfos.js");
var dbUtils = require("./models/dbUtils.js");
var res = require('./lib/resp.js');
var session = require('./lib/session.js');
var comm = require('./lib/Common.js');


var user = session.getUser();
if (!user) {
  return res.redirect('/routes/user/login.js');
}

if (request.getMethod() === 'POST') {
  var cart = session.getCart();

  var date = comm.getCnDate();
  var ordId = '' + date.getFullYear() + (date.getMonth() + 1) + date.getDate()
    + date.getHours() + date.getMinutes() + date.getSeconds() + Math.random().toString().substr(5);

  var carts = cart.items();
  if (carts.length == 0) {
    res.send("没有购买商品");
    return;
  }
  for (var i = 0; i < carts.length; i++) {
    var gs = carts[i];
    if (gs.num > 0) {
      var ordgood = {
        oid: dbUtils.makeOid(),
        ordId: ordId,
        goodsId: gs.id,
        goodsName: gs.name,
        price: gs.price,
        num: gs.num
      };
      var r = goods.decrGoodsNum(ordgood.goodsId, ordgood.num);
      if (r.rowsAffected === 0) {
        return res.send('下单失败！    ' + err.goodsName + "   商品不存在或库存不足！");
      }
      ordgoods.add(ordgood);
    }
  }
  var ordinfo = {
    oid: dbUtils.makeOid(),
    ordId: ordId,
    userId: user.oid,
    userName: user.userName,
    address: user.address,
    payType: 'RMB',
    payState: false,
    money: cart.getTotalMoney(),
    fuyan: request.formValue('fuyan'),
  };
  ordinfos.add(ordinfo);

  cart.clear();
  session.setCart(cart);
  var allData = cates.find();
  res.render('./views/flow/done.ejs', {
    ordId: ordId,
    user: user,
    cart: cart,
    tree: cates.getTree(allData),
    money: ordinfo.money,
    payForm: comm.getPayForm(ordId, ordinfo.money)
  });
}

File: routes/flow/done.go
package flow

import (
	"fmt"
	"html/template"
	"math/rand"
	"time"

	"github.com/zengming00/go-testShop/framework"
	"github.com/zengming00/go-testShop/lib"
	"github.com/zengming00/go-testShop/models"
)

func Done(ctx *framework.HandlerContext) {
	var user = ctx.GetUser()
	if user == nil {
		ctx.Redirect("/routes/user/login.go")
		return
	}
	if ctx.R.Method == "POST" {
		var cart = ctx.GetCart()

		var date = time.Now()
		var ordId = fmt.Sprintf("%.4d%.2d%.2d%.2d%.2d%.2d%d", date.Year(), int(date.Month()), date.Day(),
			date.Hour(), date.Minute(), date.Second(), rand.Uint32())

		var carts = cart.Items()
		if len(carts) == 0 {
			ctx.Send("购物车为空")
			return
		}
		for _, gs := range carts {
			if 0 < gs.Num {
				var ordgood = map[string]interface{}{
					"oid":       models.MakeOid(),
					"ordId":     ordId,
					"goodsId":   gs.Id,
					"goodsName": gs.Name,
					"price":     gs.Price,
					"num":       gs.Num,
				}
				var r = ctx.Goods.DecrGoodsNum(gs.Id, gs.Num)
				if r.RowsAffected == 0 {
					ctx.Send("下单失败！    " + gs.Name + "   商品不存在或库存不足！")
					return
				}
				ctx.Ordgoods.Add(ordgood)
			}
		}

		var money = cart.GetTotalMoney()
		var ordinfo = map[string]interface{}{
			"oid":      models.MakeOid(),
			"ordId":    ordId,
			"userId":   user.Oid,
			"userName": user.UserName,
			"address":  user.Address,
			"payType":  "RMB",
			"payState": false,
			"money":    money,
			"fuyan":    ctx.R.FormValue("fuyan"),
		}
		ctx.Ordinfos.Add(ordinfo)

		cart.Clear()
		ctx.SetCart(cart)

		var allData = ctx.Cates.Find()
		var data = map[string]interface{}{
			"tree":    ctx.Cates.GetTree(allData),
			"user":    user,
			"cart":    cart,
			"ordId":   ordId,
			"money":   money,
			"payForm": template.HTML(lib.GetPayForm(ordId, money)),
		}
		ctx.Render("./views/flow/done.html", data, lib.FuncMap)
	}
}
